Source/Core/Core/PowerPC/Jit64/Jit.cpp

```diff
@@ -1188,14 +1201,51 @@ bool Jit64::DoJit(u32 em_address, JitBlock* b, u32 nextPC)
     return false;
   }
 
-  b->codeSize = static_cast<u32>(GetCodePtr() - b->normalEntry);
-  b->originalSize = code_block.m_num_instructions;
+  return true;
+}
 
-#ifdef JIT_LOG_GENERATED_CODE
-  LogGeneratedX86(code_block.m_num_instructions, m_code_buffer, start, b);
-#endif
+void Jit64::EraseSingleBlock(const JitBlock& block)
+{
+  blocks.EraseSingleBlock(block);
+  FreeRanges();
+}
 
-  return true;
+std::vector<JitBase::MemoryStats> Jit64::GetMemoryStats() const
+{
+  return {{"near", m_free_ranges_near.get_stats()}, {"far", m_free_ranges_far.get_stats()}};
+}
+
+std::size_t Jit64::DisasmNearCode(const JitBlock& block, std::ostream& stream) const
+{
+  // The last element of the JitBlock::linkData vector is not necessarily the furthest exit.
+  // See: Jit64::JustWriteExit
+  const auto iter = std::ranges::max_element(block.linkData, {}, &JitBlock::LinkData::exitPtrs);
+
+  // Link data is not guaranteed, e.g. Jit64::WriteRfiExitDestInRSCRATCH
+  if (iter == block.linkData.end())
+    return m_disassembler->Disassemble(block.normalEntry, block.near_end, stream);
+
+  // A JitBlock's near_end only records where the XEmitter was after DoJit concludes. However, a
+  // JitBlock's exits will be modified by block linking. If Block A wants to link its final exit
+  // to the entry_point of Block B, and Block B follows Block A in memory, then the final exit's
+  // JMP will not have its destination modified but will instead be overwritten by a multibyte NOP.
+  // Trickily, Block A's near_end does not necessarily equal Block B's entry_point because Block B's
+  // entry_point is aligned to the next multiple of 4! This means the multibyte NOP may need to
+  // extend past Block A's near_end, complicating host code disassembly. If the opcode of a JMP
+  // instruction is found at the final exit, the block will be disassembled like normal. If one
+  // is not, the exit is assumed to be overwritten, and special action is taken.
```

Comment:
`near_begin` and `near_end` represent the memory region taken from the `RangeSizeSet` after `DoJit` succeeds, and their values unmodified are essential for returning that memory region back to the `RangeSizeSet`, so I doubt it.

Reply:
That sounds kind of sketchy, though. If backpatching causes a multibyte NOP to be written past the end of `near_end`, shouldn't additional memory be claimed from the RangeSizeSet at that point? I can see how it happens to work correctly in practice, but on a formal level, the JIT shouldn't be writing to memory it doesn't claim in the RangeSizeSet.

Reply:
From a cursory glance, it seems like `JitBlockCache::WriteLinkBlock` has no reasonable path to `Jit64` to get the `RangeSizeSet` it would need to modify to handle this correctly. I'm not sure what should be done about it, really.

I tracked the write past end behavior to this PR from 2015: https://github.com/dolphin-emu/dolphin/pull/2888

Reply:
Assuming the start of a block is always aligned, how about we always align the end of a block too? That way any space needed for writing a multibyte NOP will always be allocated in advance. I suppose this will lead to us disassembling a little bit of "garbage" at the end of each block, but if that's a problem, we could explicitly write INT3 instructions or something to those 0-3 bytes at the end, making it clear to the user of the disassembler that that part isn't meant to be executed.

Reply:
`AlignCode4` writes INT3 instructions as it moves the code pointer, so this suggestion works. It's a bit unfortunate that the near code instruction count when disassembling will be slightly inflated by these INT3 instructions, however.

Reply:
> It's a bit unfortunate that the near code instruction count when disassembling will be slightly inflated by these INT3 instructions, however.

Yes, I suppose that's true. But on the other hand, that space is effectively being used up by the block, in the sense that other blocks can't use it. So I think either way is okay regarding this metric.